cairo/src/runner.rs

```diff
@@ -0,0 +1,472 @@
+//! This module represents a run of a Cairo program as a consecution of execution
+//! steps, each of which define the execution logic of Cairo instructions
+
+use crate::definitions::*;
+use crate::memory::CairoMemory;
+use crate::word::{CairoWord, Decomposition};
+use ark_ff::FftField;
+
+/// A structure to store program counter, allocation pointer and frame pointer
+#[derive(Clone, Copy)]
+pub struct CairoPointers<F: FftField> {
+    /// Program counter: points to address in memory
+    pub pc: F,
+    /// Allocation pointer: points to first free space in memory
+    pub ap: F,
+    /// Frame pointer: points to the beginning of the stack in memory (for arguments)
+    pub fp: F,
+}
+
+impl<F: FftField> CairoPointers<F> {
+    /// Creates a new triple of pointers
+    pub fn new(pc: F, ap: F, fp: F) -> CairoPointers<F> {
+        CairoPointers { pc, ap, fp }
+    }
+}
+
+/// A structure to store auxiliary variables throughout computation
+pub struct CairoVariables<F: FftField> {
+    /// Destination
+    dst: Option<F>,
+    /// First operand
+    op0: Option<F>,
+    /// Second operand
+    op1: Option<F>,
+    /// Result
+    res: Option<F>,
+    /// Destination address
+    dst_addr: F,
+    /// First operand address
+    op0_addr: F,
+    /// Second operand address
+    op1_addr: F,
+    /// Size of the instruction
+    size: F,
+}
+
+impl<F: FftField> CairoVariables<F> {
+    /// This function creates an instance of a default CairoVariables struct
+    pub fn new() -> CairoVariables<F> {
+        CairoVariables {
+            dst: None,
+            op0: None,
+            op1: None,
+            res: None,
+            dst_addr: F::zero(),
+            op0_addr: F::zero(),
+            op1_addr: F::zero(),
+            size: F::zero(),
+        }
+    }
+}
+impl<F: FftField> Default for CairoVariables<F> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A data structure to store a current step of Cairo computation
+pub struct CairoStep<'a, F: FftField> {
+    /// current word of the program
+    pub mem: &'a mut CairoMemory<F>,
+    // comment instr for efficiency
+    /// current pointers
+    curr: CairoPointers<F>,
+    /// (if any) next pointers
+    next: Option<CairoPointers<F>>,
+    /// state auxiliary variables
+    vars: CairoVariables<F>,
+}
+
+impl<'a, F: FftField> CairoStep<'a, F> {
+    /// Creates a new Cairo execution step from a step index, a Cairo word, and current pointers
+    pub fn new(mem: &mut CairoMemory<F>, ptrs: CairoPointers<F>) -> CairoStep<F> {
+        CairoStep {
+            mem,
+            curr: ptrs,
+            next: None,
+            vars: CairoVariables::new(),
+        }
+    }
+
+    /// Executes a Cairo step from the current registers
+    pub fn execute(&mut self) {
+        // This order is important in order to allocate the memory in time
+        self.set_op0();
+        self.set_op1();
+        self.set_res();
+        self.set_dst();
```

Comment:
I think it'd be much nicer if the code matched the paper as closely as possible so that we can easily audit it. It also looks like CairoVariables is unnecessary here? It's just to keep track of variables between these functions right? Why not just keep track of these variables within the logic of execute?

Reply:
Oh, precisely that ordering of `op0, op1, res, dst` is the one in the paper (not sure if that's what you mean). 

Regarding the variables, do you mean we should keep those variables inside the impl of CairoStep instead of a field inside the struct itself? Answering your question, these variables only live within the step, and they do not connect to other parts of the following step, other than the logics itself. If you instead mean we could pass them as inputs/outputs of the individual functions that decompose the execution? Since there are so many of these, I thought one would get very long functions calls by doing it that way. But I am not sure this is what you meant.

Reply:
When I was reading the paper, I would have imagined the state transition to be implemented as a single function, not as a number of functions. If all these functions were inlined you wouldn't need the Variables if I understand correctly.

Reply:
Oh, in the paper it is indeed just one function... I preferred to have it as different functions because it helped me to better distinguish each of the sections of the transition function. But then sure, if I want to avoid having a Variables struct then I will need to pass these values as arguments/return. I can do it for the next version of the runner if you think it looks neater that way 👍